src/Model/Join.php

```diff
@@ -167,30 +167,39 @@ protected function init(): void
     {
         $this->_init();
 
-        // handle foreign table containing a dot
+        // owner model should have id_field set
+        $id_field = $this->getOwner()->id_field;
+        if (!$id_field) {
+            throw (new Exception('Joins owner model should have id_field set'))
+                ->addMoreInfo('model', $this->getOwner());
+        }
+
+        // handle foreign table containing a dot - that will be reverse join
         if (is_string($this->foreign_table) && strpos($this->foreign_table, '.') !== false) {
+            // split by LAST dot in foreign_table name
+            [$this->foreign_table, $this->foreign_field] = preg_split('/\.+(?=[^\.]+$)/', $this->foreign_table);
+
             if (!isset($this->reverse)) {
                 $this->reverse = true;
-                if (isset($this->master_field)) {
-                    // both master and foreign fields are set
-
-                    // master_field exists, no we will use that
-                    // if (!is_object($this->master_field) && !$this->getOwner()->hasField($this->master_field)) {
-                    throw (new Exception('You are trying to link tables on non-id fields. This is not implemented yet'))
-                        ->addMoreInfo('condition', $this->getOwner()->table . '.' . $this->master_field . ' = ' . $this->foreign_table);
-                    // } $this->reverse = 'link';
-                }
             }
+        }
 
-            // split by LAST dot in foreign_table name
-            [$this->foreign_table, $this->foreign_field] = preg_split('/\.+(?=[^\.]+$)/', $this->foreign_table);
+        if ($this->reverse === true) {
+            if (isset($this->master_field) && $this->master_field !== $id_field) {
+                throw (new Exception('You are trying to link tables on non-id fields. This is not implemented yet'))
```

Comment:
```suggestion
                throw (new Exception('Joining tables on non-id fields is not implemented yet'))
```

Reply:
I wonder what is the main issue

Reply:
If i remember correctly that was because we can't rely on fields if at least one of them is not unique id field. Something along these lines. Anyway this PR tries not to solve that.

Reply:
`we can't rely on fields if at least one of them is not unique id field.` is this important?

Reply:
All work is not done by SQL I think. At least that wasn't the case time ago. It's important in what order you insert records (reverse or not) and afterwards you have to insert respective record in other table too.

Anyway I specifically didn't address that here!
This PR is just about adding more tests and a bit _normalizing_ method how we define reverse joins. Before there was only a one and not intuitive way how to define them. And as soon as you, for example, explicitly said that join is reverse or even set master-field='id' yourself (same as default value set by ATK) it started to fail with _You are trying to link tables on non-id fields. This is not implemented yet_ exception.

Reply:
> I specifically didn't address that here

but 1/2 of the added LOC are exceptions for that, better to solve the ID limitation

Reply:
These two added expected exceptions in tests was added because previously tests was wrong. They "tested" cases which was not supported, but didn't get proper exception because of slightly wrong exception implementation in Join class (which is now fixed there).
OK I will create issue for that and see what can be done but outside of this PR.